controllers/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -49,13 +53,187 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
+				if err != nil {
+					return false, "", err
+				}
+				// new ttr created
+				return true, ttrName, nil
+			} else {
+				// something wrong happened
+				return false, "", err
+			}
+		}
+		// TODO compare TierTemplate content with TTR content
+		// if the TierTemplate has changes we need to create new TTR
+	} else {
+		// no revision was set for this TierTemplate CR, let's create a TTR for it
+		ttrName, err := r.createNewTierTemplateRevision(ctx, nsTmplTier, tierTemplate)
+		if err != nil {
+			return false, ttrName, err
+		}
+		// new ttr created
+		return true, ttrName, nil
+	}
+	// nothing changed
+	return false, "", nil
+}
+
+func (r *Reconciler) createNewTierTemplateRevision(ctx context.Context, nsTmplTier *toolchainv1alpha1.NSTemplateTier, tierTemplate *toolchainv1alpha1.TierTemplate) (string, error) {
+	ttr := NewTTR(tierTemplate, nsTmplTier)
+	ttr, err := r.createTTR(ctx, ttr, tierTemplate)
+	if err != nil {
+		// something went wrong while creating new ttr
+		return "", err
+	}
+	return ttr.GetName(), nil
+}
+
+// getNSTemplateTierRefs returns a list with all the refs from the NSTemplateTier
+func getNSTemplateTierRefs(tmplTier *toolchainv1alpha1.NSTemplateTier) []string {
+	var refs []string
+	for _, ns := range tmplTier.Spec.Namespaces {
+		refs = append(refs, ns.TemplateRef)
+	}
+	if tmplTier.Spec.ClusterResources != nil {
+		refs = append(refs, tmplTier.Spec.ClusterResources.TemplateRef)
+	}
+
+	roles := make([]string, 0, len(tmplTier.Spec.SpaceRoles))
+	for r := range tmplTier.Spec.SpaceRoles {
+		roles = append(roles, r)
+	}
+	for _, r := range roles {
+		refs = append(refs, tmplTier.Spec.SpaceRoles[r].TemplateRef)
+	}
+	return refs
+}
+
+func (r *Reconciler) createTTR(ctx context.Context, ttr *toolchainv1alpha1.TierTemplateRevision, tmplTier *toolchainv1alpha1.TierTemplate) (*toolchainv1alpha1.TierTemplateRevision, error) {
+	err := r.Client.Create(ctx, ttr)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("unable to create TierTemplateRevision: %w", err)
+	}
+
+	logger := log.FromContext(ctx)
+	logger.Info("created TierTemplateRevision", "tierTemplateRevision.Name", ttr.Name, "tierTemplate.Name", tmplTier.Name)
+	return ttr, nil
+}
+
+// NewTTR creates a TierTemplateRevision CR for a given TierTemplate object.
+func NewTTR(tierTmpl *toolchainv1alpha1.TierTemplate, nsTmplTier *toolchainv1alpha1.NSTemplateTier) *toolchainv1alpha1.TierTemplateRevision {
+	tierName := nsTmplTier.GetName()
+	tierTemplateName := tierTmpl.GetName()
+	labels := map[string]string{
+		toolchainv1alpha1.TierLabelKey:        tierName,
+		toolchainv1alpha1.TemplateRefLabelKey: tierTemplateName,
+	}
+
+	newTTRName := fmt.Sprintf("%s-%s-", tierName, tierTemplateName)
+	// we have to cut it down to the max allowed length
+	// 6 = 5 is generateName suffix length + 1 the "-" char
+	mexNameLength := validation.DNS1123LabelMaxLength + 6
+	if len(newTTRName) > mexNameLength {
+		newTTRName = newTTRName[0:mexNameLength]
+	}
```

Comment:
two things
1. it should be `validation.DNS1123LabelMaxLength - 6` to not exceed the `validation.DNS1123LabelMaxLength` when there is the 6 chars suffix
2. there is a typo in `mexNameLength`

```suggestion
	maxNameLength := validation.DNS1123LabelMaxLength - 6
	if len(newTTRName) > maxNameLength {
		newTTRName = newTTRName[0:maxNameLength]
	}
```

Reply:
ohhh jeez 🤦‍♂️ , your right! great catch! Thanks for fixing it!